File: src/main/java/com/lessmarkup/userinterface/nodehandlers/common/PropertiesNodeHandler.java
package com.lessmarkup.userinterface.nodehandlers.common;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.lessmarkup.framework.helpers.*;
import com.lessmarkup.framework.nodehandlers.AbstractNodeHandler;
import com.lessmarkup.interfaces.module.ModuleConfiguration;
import com.lessmarkup.interfaces.module.ModuleProvider;
import com.lessmarkup.interfaces.recordmodel.InputFieldType;
import com.lessmarkup.interfaces.structure.Property;

import java.net.URL;
import java.util.*;

public abstract class PropertiesNodeHandler extends AbstractNodeHandler {

    private final ModuleProvider moduleProvider;
    private final List<JsonObject> properties = new ArrayList<>();

    protected PropertiesNodeHandler(ModuleProvider moduleProvider) {
        this.moduleProvider = moduleProvider;
    }

    protected void addProperty(String name, InputFieldType type, String value) {
        JsonObject model = new JsonObject();
        model.addProperty("name", name);
        model.addProperty("type", type.toString());
        model.addProperty("value", value);
        properties.add(model);
    }

    @Override
    public String getViewType() {
        return "properties";
    }

    @Override
    public JsonObject getViewData() {

        URL moduleUrl = getClass().getProtectionDomain().getCodeSource().getLocation();

        Optional<ModuleConfiguration> moduleConfiguration = moduleProvider.getModules().stream().filter(m -> m.getUrl().equals(moduleUrl)).findFirst();

        JsonObject ret = new JsonObject();
        JsonArray propertiesArray = new JsonArray();
        
        for (PropertyDescriptor property : TypeHelper.getProperties(getClass())) {
            Property propertyAttribute = property.getAnnotation(Property.class);

            if (propertyAttribute == null) {
                continue;
            }

            Object value = property.getValue(this);

            if (value == null) {
                continue;
            }

            JsonObject model = new JsonObject();
            model.addProperty("name", LanguageHelper.getFullTextId(moduleConfiguration.get().getModuleType(), propertyAttribute.textId()));
            model.addProperty("type", propertyAttribute.type().toString());
            model.addProperty("value", value.toString());
            
            switch (propertyAttribute.type()) {
                case IMAGE:
                    long imageId = Long.parseLong(value.toString());
                    model.addProperty("value", ImageHelper.getImageUrl(imageId));
                    break;
            }

            propertiesArray.add(model);
        }

        for (JsonObject item : this.properties) {
            propertiesArray.add(item);
        }

        ret.add("properties", propertiesArray);

        return ret;
    }
}

File: src/main/scala/com/lessmarkup/userinterface/nodehandlers/common/PropertiesNodeHandler.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.nodehandlers.common

import java.net.URL

import com.google.gson.{JsonArray, JsonObject}
import com.lessmarkup.framework.helpers._
import com.lessmarkup.framework.nodehandlers.{AbstractNodeHandler, NodeHandlerConfiguration}
import com.lessmarkup.interfaces.annotations.{InputFieldType, Property}
import com.lessmarkup.interfaces.module.{ModuleConfiguration, ModuleProvider}

object PropertiesNodeHandler {
  class Property(val name: String, val fieldType: InputFieldType, val value: String)
}

abstract class PropertiesNodeHandler(moduleProvider: ModuleProvider, configuration: NodeHandlerConfiguration, propertiesSource: Seq[PropertiesNodeHandler.Property])
  extends AbstractNodeHandler(configuration) {

  private final val properties: Seq[JsonObject] = propertiesSource.map(p => {
    val model: JsonObject = new JsonObject
    model.addProperty("name", p.name)
    model.addProperty("type", p.fieldType.toString)
    model.addProperty("value", p.value)
    model
  })

  override def getViewType = "properties"

  override def getViewData: Option[JsonObject] = {
    val moduleUrl: URL = getClass.getProtectionDomain.getCodeSource.getLocation
    val moduleConfiguration: Option[ModuleConfiguration] = moduleProvider.getModules.find(m => m.getUrl.equals(moduleUrl))
    val ret: JsonObject = new JsonObject
    val propertiesArray: JsonArray = new JsonArray

    for (
      property <- TypeHelper.getProperties(getClass);
      attribute = property.getAnnotation(classOf[Property]);
      value = property.getValue(this)
    ) {
      val model: JsonObject = new JsonObject
      model.addProperty("name", LanguageHelper.getFullTextId(moduleConfiguration.get.getModuleType, attribute.textId))
      model.addProperty("type", attribute.`type`.toString)
      model.addProperty("value", value.toString)

      if (attribute.`type`() == InputFieldType.IMAGE) {
        val imageId = value.toString.toLong
        model.addProperty("value", ImageHelper.getImageUrl(imageId))
      }

      propertiesArray.add(model)
    }

    properties.foreach(p => propertiesArray.add(p))

    ret.add("properties", propertiesArray)

    Option(ret)
  }
}